Confusing diagnostic when `use Self::Variant` inside an impl

I tried this ([play](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021)):

```rust
pub enum Uwu {
    Hi
}

impl Uwu {
    fn a() -> Uwu {
        use Self::Hi;
        Hi
    }
}
```


Heres the current output:
```
   Compiling playground v0.0.1 (/playground)
error[E0432]: unresolved import `Self`
 --> src/lib.rs:8:13
  |
8 |         use Self::Hi;
  |             ^^^^ `Self` is only available in impls, traits, and type definitions

For more information about this error, try `rustc --explain E0432`.
error: could not compile `playground` due to previous error
```

The label that says "`Self` is only available in impls" is misleading since it is already in an impl.

--- a/compiler/rustc_resolve/src/diagnostics.rs
+++ b/compiler/rustc_resolve/src/diagnostics.rs
@@ -1832,7 +1832,17 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
             }
             (msg, None)
         } else if ident.name == kw::SelfUpper {
-            ("`Self` is only available in impls, traits, and type definitions".to_string(), None)
+            // As mentioned above, `opt_ns` being `None` indicates a module path in import.
+            // We can use this to improve a confusing error for, e.g. `use Self::Variant` in an
+            // impl
+            if opt_ns.is_none() {
+                ("`Self` cannot be used in imports".to_string(), None)
+            } else {
+                (
+                    "`Self` is only available in impls, traits, and type definitions".to_string(),
+                    None,
+                )
+            }
         } else if ident.name.as_str().chars().next().map_or(false, |c| c.is_ascii_uppercase()) {
             // Check whether the name refers to an item in the value namespace.
             let binding = if let Some(ribs) = ribs {

--- a/tests/ui/use/use-self-type.stderr
+++ b/tests/ui/use/use-self-type.stderr
@@ -8,7 +8,7 @@ error[E0432]: unresolved import `Self`
   --> $DIR/use-self-type.rs:6:13
    |
 LL |         use Self::f;
-   |             ^^^^ `Self` is only available in impls, traits, and type definitions
+   |             ^^^^ `Self` cannot be used in imports
 
 error: aborting due to 2 previous errors
 

